build/image_build.go

```diff
@@ -0,0 +1,117 @@
+package build
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/jhoonb/archivex"
+	"github.com/otiai10/copy"
+)
+
+func BuildImage(imageName, rootDir string) error {
+	// Make temp directory
+	tmpDir, err := ioutil.TempDir("", "nnpd-")
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(tmpDir+"/var/lib/nnpd", os.ModePerm)
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	// Copy rootDir into temp/var/lib/nnpd
+	opt := copy.Options{
+		Skip: func(src string) (bool, error) {
+			return strings.HasSuffix(src, ".git"), nil
+		},
+	}
+	err = copy.Copy(rootDir, tmpDir+"/var/lib/nnpd", opt)
+	if err != nil {
+		return err
+	}
+
+	// Change directory to tmpDir
+	err = os.Chdir(tmpDir)
+	if err != nil {
+		return err
+	}
+
+	// Tar temp/var/lib/nnpd
+	tarDir(tmpDir+"/nnpd.tar", "var/lib/nnpd")
+
+	// Write Dockerfile
+	d1 := []byte("FROM ubuntu:20.04\nCOPY ./nnpd.tar /tmp/nnpd.tar\nENTRYPOINT tar -xvf /tmp/nnpd.tar -C /alloc >/dev/null 2>&1")
+	err = ioutil.WriteFile(tmpDir+"/Dockerfile", d1, 0744)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	cli, err := client.NewEnvClient()
+	if err != nil {
+		return err
+	}
+
+	file, err := os.OpenFile(tmpDir+"/nnpd.tar", os.O_RDWR, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	fileInfo, err := os.Stat(tmpDir+"/nnpd.tar")
+	if err != nil {
+		return err
+	}
```

Comment:
Can we move (modularize) `os.OpenFile()` and `os.Stat()` into a separate function. This is being duplicated `twice` here.

Reply:
Did you want one function which calls `os.OpenFile` and `os.Stat()`?

Reply:
Yeah, so you can call that for both `nnpd.tar` and `Dockerfile`. Modularize and reuse!

Reply:
got it